frontend/src/components/AppHeader.tsx

```diff
@@ -1,53 +1,57 @@
-import { useCallback, useContext, useEffect, useRef, useState } from "react";
-import { UserContext } from "@/lib/UserContext";
+import { useCallback, useEffect, useRef, useState } from "react";
 import Link from "next/link";
 import Image from "next/image";
-import AppNavigation from "./AppNavigation";
-
 import styles from "@/styles/Nav.module.css";
-
+import AppNavigation from "./AppNavigation";
+import { useUser } from "@/context/UserContext";
+import { useWeb3 } from "@/context/Web3Context";
 import { magic } from "@/lib/magic";
-import { getUserData } from "@/lib/utils";
 
 export default function AppHeader({}) {
   const [navbarOpen, setNavbarOpen] = useState(false);
-  const [user, setUser] = useContext(UserContext);
+  const { user, setUser } = useUser();
+  const { initializeWeb3 } = useWeb3();
+
+  // Reference to the navigation menu
   const navMenuRef = useRef(null);
 
-  function openWallet() {
-    magic.wallet.getInfo().then((walletInfo) => {
-      if (walletInfo?.walletType == "magic") {
-        // NOTE: this will only work if the user has connected via a
-        // magic wallet, not via browser wallet (e.g. MetaMask)
-        magic.wallet.showUI().catch((err) => console.error(err));
-      } else {
-        // for non-magic wallets, copy the full wallet address to the clipboard
-        navigator.clipboard
-          .writeText(user?.address)
-          .then((res) =>
-            alert(`ETH wallet address copied to clipboard: ${user?.address}`),
-          );
-      }
-    });
-  }
-
-  function disconnect() {
-    // disconnect from magic
-    magic.wallet.disconnect();
-
-    // clear the state
-    setUser({});
-  }
-
-  function loginWithConnect() {
-    magic.wallet
-      .connectWithUI()
-      .then((res) => {
-        getUserData().then((data) => setUser(data));
-      })
-      .catch((err) => console.error(err));
-  }
+  // Function to handle wallet opening
+  const openWallet = async () => {
+    try {
+      // Try to show the magic wallet UI
+      // This will only work if the user has connected via a magic wallet, not via browser wallet (e.g. MetaMask)
+      await magic.wallet.showUI();
+    } catch (error) {
+      console.error("openWallet", error);
+      // for non-magic wallets, copy the full wallet address to the clipboard
+      await navigator.clipboard.writeText(user?.address);
+      alert(`ETH wallet address copied to clipboard: ${user?.address}`);
+    }
+  };
+
+  // Function to handle disconnection
+  const disconnect = async () => {
+    // Disconnect from magic
+    await magic.user.logout();
+
+    // Clear the user state
+    setUser(null);
```

Comment:
After a user disconnects, are you re-initializing the web3 instance? Thinking of the case where a user logs in with metamask, then disconnects, the web3 provider will be metamask even though they disconnected - any non-logged in state should have the web3 provider be the magic.rpcProvider (that will be returned by magic.wallet.getProvider() when there's no user logged in)

Reply:
Another good call. Will fix